Resolve tweaks

nnethercote opened the pull request:
A couple of small perf improvements, and some minor refactorings, all in `rustc_resolve`.

r? @petrochenkov

nnethercote committed:
Use iteration instead of indexing to access ribs.

This gives a small but noticeable performance improvement.

nnethercote committed:
Optimize `check_keyword_case`.

`to_lowercase` allocates, but `eq_ignore_ascii_case` doesn't. This path
is hot enough that this makes a small but noticeable difference in
benchmarking.

nnethercote committed:
Tweak a `resolutions` loop.

In this case field access is more concise and easier to read than
destructuring, and it matches how other similar loops are done
elsewhere.

nnethercote committed:
Use `for_each_child` in a suitable place.

`for_each_child` exists for this exact pattern.

nnethercote committed:
Use an atom comparison for a keyword check.

Instead of a string comparison.

nnethercote committed:
Remove two `_ext` methods.

`resolve_ident_in_module` is a very thin wrapper around
`resolve_ident_in_module_ext`, and `resolve_ident_in_module_unadjusted`
is a very thin wrapper around `resolve_ident_in_module_unadjusted_ext`.
The wrappers make the call sites slightly more concise, but I don't
think that's worth the extra code and indirection.

This commit removes the two wrappers and removes the `_ext` suffixes
from the inner methods.

nnethercote committed:
Move a call outside a loop.

This path isn't hot enough for this to affect performance, but there's
no point repeating the same computation multiple times.

nnethercote commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 8e22596659c3a920ace15d4da9a5bc1734cc8374 with merge 4e79a44c027ccc8debd841461a54f63cac112dfb...
<!-- homu: {"type":"TryBuildStarted","head_sha":"8e22596659c3a920ace15d4da9a5bc1734cc8374","merge_sha":"4e79a44c027ccc8debd841461a54f63cac112dfb"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/11737711438/job/32702878648)
Build commit: 4e79a44c027ccc8debd841461a54f63cac112dfb (`4e79a44c027ccc8debd841461a54f63cac112dfb`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/11737711438/job/32702878648"},"merge_sha":"4e79a44c027ccc8debd841461a54f63cac112dfb"} -->

rust-timer commented:
Queued 4e79a44c027ccc8debd841461a54f63cac112dfb with parent 78bb5ee79e0261e8e47476b631da02acc1cb03ef, future [comparison URL](https://perf.rust-lang.org/compare.html?start=78bb5ee79e0261e8e47476b631da02acc1cb03ef&end=4e79a44c027ccc8debd841461a54f63cac112dfb).
There are currently 0 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.2 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (4e79a44c027ccc8debd841461a54f63cac112dfb): [comparison URL](https://perf.rust-lang.org/compare.html?start=78bb5ee79e0261e8e47476b631da02acc1cb03ef&end=4e79a44c027ccc8debd841461a54f63cac112dfb&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=78bb5ee79e0261e8e47476b631da02acc1cb03ef&end=4e79a44c027ccc8debd841461a54f63cac112dfb&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.3% | [-0.5%, -0.1%] | 40    |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.3% | [-0.5%, -0.1%] | 40    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=78bb5ee79e0261e8e47476b631da02acc1cb03ef&end=4e79a44c027ccc8debd841461a54f63cac112dfb&stat=max-rss)
<details>
<summary>Results (primary 1.3%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 3.3%  | [1.0%, 5.6%]   | 2     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.7% | [-0.8%, -0.7%] | 2     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 1.3%  | [-0.8%, 5.6%]  | 4     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=78bb5ee79e0261e8e47476b631da02acc1cb03ef&end=4e79a44c027ccc8debd841461a54f63cac112dfb&stat=cycles:u)
<details>
<summary>Results (secondary 7.9%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 7.9% | [7.9%, 7.9%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=78bb5ee79e0261e8e47476b631da02acc1cb03ef&end=4e79a44c027ccc8debd841461a54f63cac112dfb&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 780.076s -> 781.343s (0.16%)
**Artifact size**: 335.37 MiB -> 335.36 MiB (-0.00%)

nnethercote commented:
@petrochenkov: I was going to request a different reviewer because (a) I know you are busy, and (b) none of these changes involve any resolver-specific expertise. But you can review if you want.

jieyouxu commented:
I can take a look.

r? jieyouxu

jieyouxu commented on compiler/rustc_resolve/src/ident.rs:
```diff
@@ -814,34 +793,10 @@ impl<'ra, 'tcx> Resolver<'ra, 'tcx> {
     /// Invariant: if `finalize` is `Some`, expansion and import resolution must be complete.
     #[instrument(level = "debug", skip(self))]
-    fn resolve_ident_in_module_unadjusted_ext(
+    fn resolve_ident_in_module_unadjusted(
```
Suggestion: github UI doesn't let me comment outside the diff, but should we introduce an enum for `RestrictedShadowing`? The `false` looks very uninformative and it is not documented.

nnethercote replied:
I'll do an enum.

jieyouxu reviewed (approved):
Thanks, one tiny nit regarding a `bool` param for a fn that has multiple params that looks confusing at call-sites. But otherwise, this looks like a nice cleanup!

jieyouxu commented:
@rustbot author

petrochenkov commented:
>But you can review if you want.

Yes, I self assign if I want to review, or at least take a look.

jieyouxu commented:
(My bad I didn't notice it)

petrochenkov commented on compiler/rustc_resolve/src/lib.rs:
```diff
@@ -534,8 +534,8 @@ struct BindingKey {
-    /// 0 if ident is not `_`, otherwise a value that's unique to the specific
-    /// `_` in the expanded AST that introduced this binding.
+    /// 0 if ident is not `_` or ``, otherwise a value that's unique to the specific
+    /// `_`/`` in the expanded AST that introduced this binding.
```
https://github.com/rust-lang/rust/pull/115367 was reverted in https://github.com/rust-lang/rust/pull/131045, and even if it weren't `kw::Empty` shouldn't have been used for this anyway.

nnethercote replied:
Huh, interesting. `empty_disambiguator` was added in #115367 but wasn't removed by the revert in #131045. @compiler-errors, you did the revert, do you remember how you went about it -- was it a straightforward reversion? Was leaving `empty_disambiguator` behind an oversight?

There is one test (`tests/ui/dollar-crate/dollar-crate-is-keyword.rs`) that hits the empty identifier path. But if I remove `empty_disambiguator` all tests pass. So I've removed the commit changing this comment and added a new commit that removes `empty_disambiguator`.

nnethercote committed:
Replace the `restricted_shadowing` boolean argument with an enum.

It makes the code clearer.

nnethercote committed:
Remove `Resolver::empty_disambiguator`.

It was added in #115367 for anonymous ADTs. Those changes were then
reverted in #131045, but `empty_disambiguator` was left behind, perhaps
by mistake. It seems to be unnecessary.

nnethercote commented:
I have updated. I removed one old commit (the one updating the comment on `BindingKey::disambiguator`) and added two new commits (one adding the `Shadowing` enum, and one removing `Resolver::empty_disambiguator`).

nnethercote commented:
@rustbot ready

petrochenkov commented:
@bors r+

bors commented:
:pushpin: Commit 12747f188a9dba7b13cf1ccb01fef74c6ce3294e has been approved by `petrochenkov`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=petrochenkov 12747f188a9dba7b13cf1ccb01fef74c6ce3294e -->
<!-- homu: {"type":"Approved","sha":"12747f188a9dba7b13cf1ccb01fef74c6ce3294e","approver":"petrochenkov","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 12747f188a9dba7b13cf1ccb01fef74c6ce3294e with merge ee612c45f00391aff71ec0c52b7fc08fae18c711...
<!-- homu: {"type":"BuildStarted","head_sha":"12747f188a9dba7b13cf1ccb01fef74c6ce3294e","merge_sha":"ee612c45f00391aff71ec0c52b7fc08fae18c711"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/11920697857/job/33231456551)
Approved by: petrochenkov
Pushing ee612c45f00391aff71ec0c52b7fc08fae18c711 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"petrochenkov","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/11920697857/job/33231456551"},"merge_sha":"ee612c45f00391aff71ec0c52b7fc08fae18c711"} -->

rust-timer commented:
Finished benchmarking commit (ee612c45f00391aff71ec0c52b7fc08fae18c711): [comparison URL](https://perf.rust-lang.org/compare.html?start=78993684f2ba22082197d63627f3e91069911839&end=ee612c45f00391aff71ec0c52b7fc08fae18c711&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=78993684f2ba22082197d63627f3e91069911839&end=ee612c45f00391aff71ec0c52b7fc08fae18c711&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.3% | [-0.6%, -0.1%] | 68    |
| Improvements ✅ <br /> (secondary) | -1.1% | [-1.1%, -1.1%] | 1     |
| All ❌✅ (primary)                 | -0.3% | [-0.6%, -0.1%] | 68    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=78993684f2ba22082197d63627f3e91069911839&end=ee612c45f00391aff71ec0c52b7fc08fae18c711&stat=max-rss)
<details>
<summary>Results (primary -0.5%, secondary 3.0%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.9%  | [0.9%, 0.9%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 3.0%  | [3.0%, 3.0%]   | 1     |
| Improvements ✅ <br /> (primary)   | -1.9% | [-1.9%, -1.9%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.5% | [-1.9%, 0.9%]  | 2     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=78993684f2ba22082197d63627f3e91069911839&end=ee612c45f00391aff71ec0c52b7fc08fae18c711&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=78993684f2ba22082197d63627f3e91069911839&end=ee612c45f00391aff71ec0c52b7fc08fae18c711&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 793.886s -> 795.801s (0.24%)
**Artifact size**: 335.28 MiB -> 335.27 MiB (-0.00%)
